### PR TITLE
ref(py3): coerce map, filter, zip builtins into lists and enforce this as a lint rule

### PR DESCRIPTION
**Pull request by joshuarli**
TODO: rebase against origin/master once https://github.com/getsentry/sentry/pull/17120 lands

Also blocked by https://github.com/getsentry/sentry-flake8/pull/8 + new release (then would need to bump travis and requirements here)

The map, filter, and zip builtins are iterables under Python 3, so in order to ensure compatibilty these need to all be cast into lists. There are probably some places where they're exclusively used as an iterable i.e. consumed, but I don't think it's worth the time to audit.

```python
import sys
listcast = len(sys.argv) > 1

data = list(range(100))
for _ in range(100000):
    if listcast:
        list(map(lambda _:_, data))
    else:
        map(lambda _:_, data)
```
```
python2 bench.py  1.07s user 0.02s system 100% cpu 1.086 total
python2 bench.py list  1.05s user 0.04s system 100% cpu 1.085 total
python3 bench.py  0.08s user 0.00s system 99% cpu 0.081 total
python3 bench.py list  0.91s user 0.01s system 99% cpu 0.920 total
```

From my naive testing, I observed no noticeable difference in speed. I also snuck in a print(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss) at the end, and saw no clearly increased memory usage.

### CODE REVIEW

File: src/sentry/api/endpoints/group_similar_issues.py [hunk 1]
```diff
@@ -17,10 +19,10 @@ def get(self, request, group):
         if limit is not None:
             limit = int(limit) + 1  # the target group will always be included
 
-        results = filter(
+        results = list(filter(
```
Comment (wedamija) [inline src/sentry/api/endpoints/group_similar_issues.py]: Why do we need the `list` here, isn't it built into the compat layer?

File: src/sentry/api/paginator.py [hunk 2]
```diff
@@ -10,6 +10,8 @@
 from django.utils import timezone
 
 from sentry.utils.cursors import build_cursor, Cursor, CursorResult
+from sentry.utils.compat import map
+from sentry.utils.compat import zip
```
Comment (wedamija) [inline src/sentry/api/paginator.py]: Probably could be same import

File: src/sentry/deletions/base.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def filter_relations(self, child_relations):
             return child_relations
 
         return list(
```
Comment (wedamija) [inline src/sentry/deletions/base.py]: List is redundant here

File: src/sentry/event_manager.py [hunk 4]
```diff
@@ -405,7 +406,7 @@ def should_filter(self):
             return (True, FilterStatKeys.ERROR_MESSAGE)
 
         for exc in get_path(self._data, "exception", "values", filter=True, default=[]):
-            message = u": ".join(filter(None, map(exc.get, ["type", "value"])))
+            message = u": ".join([_f for _f in map(exc.get, ["type", "value"]) if _f])
```
Comment (wedamija) [inline src/sentry/event_manager.py]: `_f is not None` maps to the original meaning

File: src/sentry/eventstore/models.py [hunk 5]
```diff
@@ -338,9 +339,7 @@ def get_hashes(self, force_config=None):
             if hashes is not None:
                 return hashes
 
-        return filter(
-            None, [x.get_hash() for x in self.get_grouping_variants(force_config).values()]
-        )
+        return [_f for _f in [x.get_hash() for x in self.get_grouping_variants(force_config).values()] if _f]
```
Comment (wedamija) [inline src/sentry/eventstore/models.py]: `_f is not None`

File: src/sentry/grouping/enhancer.py [hunk 6]
```diff
@@ -147,7 +148,7 @@ def matches_frame(self, frame_data, platform):
 
     def _to_config_structure(self):
         if self.key == "family":
-            arg = "".join(filter(None, [FAMILIES.get(x) for x in self.pattern.split(",")]))
+            arg = "".join([_f for _f in [FAMILIES.get(x) for x in self.pattern.split(",")] if _f])
```
Comment (wedamija) [inline src/sentry/grouping/enhancer.py]: I think this could just be one list comprehension (same for lower in this file)

**Review by wedamija — COMMENTED** [review]
A few recurring issues:

 - Wrapping uses of your compat `list`, `map`, `zip` methods in list, even though they already did that
 - Converting `filter(None, <list>)` to [a for a in <list> if a]`. Should be `if a is not None` to retain original meaning

**Comment by joshuarli** [comment]
> A few recurring issues:
> 
>     * Wrapping uses of your compat `list`, `map`, `zip` methods in list, even though they already did that
> 
>     * Converting `filter(None, <list>)` to [a for a in  if a]`. Should be `if a is not None` to retain original meaning

Yep, that's the behavior of modernize. My compat methods obsolete the need to wrap in lists, I'm going to go through and remove all those casts. It was useful for putting those imports in place though.

Regarding the filter thing, good catch. Honestly not sure why it was replaced with a truthy check, I'll manually convert those to explicit `is not None`.

**Comment by wedamija** [comment]
> > A few recurring issues:
> > ```
> > * Wrapping uses of your compat `list`, `map`, `zip` methods in list, even though they already did that
> > 
> > * Converting `filter(None, <list>)` to [a for a in  if a]`. Should be `if a is not None` to retain original meaning
> > ```
> 
> Yep, that's the behavior of modernize. My compat methods obsolete the need to wrap in lists, I'm going to go through and remove all those casts. It was useful for putting those imports in place though.
> 
> Regarding the filter thing, good catch. Honestly not sure why it was replaced with a truthy check, I'll manually convert those to explicit `is not None`.

Damn, looks like I'm wrong: https://docs.python.org/2/library/functions.html#filter. It's actually just doing a truthy check!

**Comment by joshuarli** [comment]
> > > A few recurring issues:
> > > ```
> > > * Wrapping uses of your compat `list`, `map`, `zip` methods in list, even though they already did that
> > > 
> > > * Converting `filter(None, <list>)` to [a for a in  if a]`. Should be `if a is not None` to retain original meaning
> > > ```
> > 
> > 
> > Yep, that's the behavior of modernize. My compat methods obsolete the need to wrap in lists, I'm going to go through and remove all those casts. It was useful for putting those imports in place though.
> > Regarding the filter thing, good catch. Honestly not sure why it was replaced with a truthy check, I'll manually convert those to explicit `is not None`.
> 
> Damn, looks like I'm wrong: https://docs.python.org/2/library/functions.html#filter. It's actually just doing a truthy check!

Ah, thanks for actually searching the docs for that haha.

**Comment by joshuarli** [comment]
Damn, there's a lot of `list` casts. Honestly I'm just gonna fork modernize and get it to 1) not cast and 2) replace with our import.

**Comment by joshuarli** [comment]
@wedamija Superseded by https://github.com/getsentry/sentry/pull/17270.

**Pull request closed without merging**